### PR TITLE
Better error logging

### LINKED ISSUE #982: More informative error messages in prerendering
**Is your feature request related to a problem? Please describe.**
If a page errors during prerendering, you see something like this in your terminal:

```
500 /foo
```

This isn't very helpful for diagnosing the underlying error. If it happens locally, you can `svelte-kit start` and navigate to `/foo` and see what happened, but if it happens in CI it's a pain (and sometimes errors _only_ happen in CI).

**Describe the solution you'd like**
It would be helpful if this code... https://github.com/sveltejs/kit/blob/ca108a672a009b9937f9955d81fffca7f3d8c970/packages/kit/src/core/adapt/prerender.js#L90-L105 ...somehow gave access to the underlying error. It probably shouldn't be on the response directly, but it could maybe be passed back via a side-channel, a la `dependencies`.

**Describe alternatives you've considered**
None

**How important is this feature to you?**
Nice-to-have

### PR DESCRIPTION
**Pull request by Rich-Harris**
This started out as a PR to address #982 specifically, but I ended up fixing a bunch of adjacent code:

* the `SSRRenderOptions` type is much neater, with the options passed to `app.render` separated out
* added `SSRRenderState` as distinct from `SSRRenderOptions` — this is used for tracking prerender dependencies and preventing infinite fetch loops. It means that the built app only needs to create the render options object once
* exceptions are always console-logged, whether from endpoints or pages. This fixes #982 without adding any additional machinery
* error logging is more readable in dev
* error responses (whether explicit or triggered by exceptions) in prerender will show which page linked to the erroring page, or which page fetched the erroring endpoint. This makes it much easier to find broken links etc

### Before submitting the PR, please make sure you do the following
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpx changeset` and following the prompts

### CODE REVIEW

File: packages/kit/src/core/adapt/prerender.js [hunk 1]
```diff
@@ -86,8 +87,11 @@ export async function prerender({ cwd, out, log, config, build_data, force }) {
 		}
 	});
 
-	/** @param {string} path */
-	async function visit(path) {
+	/**
+	 * @param {string} path
+	 * @param {string} parent
+	 */
+	async function visit(path, parent) {
```
Comment (benmccann) [inline packages/kit/src/core/adapt/prerender.js:94]: `parent` looks unused

Comment (benmccann) [inline packages/kit/src/core/adapt/prerender.js:94]: `parent_path` might be a clearer name? Or could update the JSDoc to clarify that it's a path

**Comment by benmccann** [comment]
Cool!! I wonder if there are any tests we should add for this

File: packages/kit/src/core/adapt/prerender.js (hunk 1 above)
Comment (Rich-Harris) [reply packages/kit/src/core/adapt/prerender.js:94]: I feel like it's obvious from usage. If you go by the name alone then you might think that the `parent_path` of `/foo/bar` was `/foo`, whereas if you scan down to where it's used, it becomes clear what it does

**Comment by Rich-Harris** [comment]
> Cool!! I wonder if there are any tests we should add for this

Prerendering tests in general are a bit of a TODO

**Comment by datio** [comment]
There's a [regression](https://discord.com/channels/457912077277855764/819723698415599626/868119360793886720) when a unicode path is `visit`'ed because of the following change:

https://github.com/sveltejs/kit/blame/80f5b767d75be6634c36510aadd604014f76a25b/packages/kit/src/core/adapt/prerender.js#L196

If it's changed, say, to:
```javascript
await visit(decodeURI(path), parsed.pathname.replace(config.kit.paths.base, ''));
```
It builds correctly.

**Comment by benmccann** [comment]
@datio can you file a new issue instead of commenting on a closed PR? this will for sure get lost otherwise

**Pull request merged**